File: android/src/main/kotlin/io/github/newbalancem5/yandex_login_sdk/YandexLoginSdkPlugin.kt
package io.github.newbalancem5.yandex_login_sdk

import androidx.activity.result.ActivityResultLauncher
import androidx.fragment.app.FragmentActivity
import com.yandex.authsdk.YandexAuthException
import com.yandex.authsdk.YandexAuthLoginOptions
import com.yandex.authsdk.YandexAuthOptions
import com.yandex.authsdk.YandexAuthResult
import com.yandex.authsdk.YandexAuthSdk
import com.yandex.authsdk.internal.strategy.LoginType
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class YandexLoginSdkPlugin :
    FlutterPlugin,
    ActivityAware,
    MethodCallHandler {

    private lateinit var channel: MethodChannel
    private var activity: FragmentActivity? = null
    private var sdk: YandexAuthSdk? = null
    private var launcher: ActivityResultLauncher<YandexAuthLoginOptions>? = null
    private var pendingResult: Result? = null
    private var loggingEnabled = false

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        val act = binding.activity as? FragmentActivity
            ?: throw IllegalStateException(
                "yandex_login_sdk requires the host Activity to be a FragmentActivity " +
                    "(use FlutterFragmentActivity in MainActivity.kt)."
            )
        activity = act
        recreateSdk(act)
    }

    /**
     * (Re)creates the SDK and the result launcher for [act]. REGISTRY_KEY is
     * stable across Activity recreations, so a result produced while the
     * Activity was being recreated (e.g. rotation during sign-in) is
     * redelivered here by the restored registry.
     */
    private fun recreateSdk(act: FragmentActivity) {
        launcher?.unregister()
        sdk = YandexAuthSdk.create(YandexAuthOptions(act, loggingEnabled))
        launcher = act.activityResultRegistry.register(
            REGISTRY_KEY,
            sdk!!.contract,
        ) { result -> deliverResult(result) }
    }

    override fun onDetachedFromActivity() {
        disposeActivity()
        pendingResult?.error("DETACHED", "Activity detached during sign-in", null)
        pendingResult = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) =
        onAttachedToActivity(binding)

    override fun onDetachedFromActivityForConfigChanges() {
        // Keep pendingResult: the auth result survives the config change and
        // is redelivered through the re-registered launcher.
        disposeActivity()
    }

    private fun disposeActivity() {
        launcher?.unregister()
        launcher = null
        sdk = null
        activity = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "signIn" -> handleSignIn(call, result)
            "signOut" -> handleSignOut(result)
            else -> result.notImplemented()
        }
    }

    /**
     * Clears local sign-in state.
     *
     * The Yandex `authsdk` (3.x) is stateless and exposes no logout/revoke, so
     * there is no native SDK session to clear. This resolves successfully as a
     * documented no-op; the app should drop its own copy of the token. It does
     * not revoke the token on Yandex's servers nor clear the Yandex-app /
     * Chrome-tab cookie session.
     */
    private fun handleSignOut(result: Result) {
        nativeLog("debug", "signOut(): no-op on Android (stateless authsdk)")
        result.success(null)
    }

    private fun handleSignIn(call: MethodCall, result: Result) {
        if (pendingResult != null) {
            result.error("BUSY", "Another sign-in is in progress", null)
            return
        }
        if (launcher == null) {
            result.error("NO_ACTIVITY", "Plugin is not attached to a FragmentActivity", null)
            return
        }
        val nativeLogging = call.argument<Boolean>("nativeLogging") ?: false
        if (nativeLogging != loggingEnabled) {
            loggingEnabled = nativeLogging
            activity?.let { recreateSdk(it) }
        }
        // authsdk 3.2+ lets the login options override the manifest clientId
        // at runtime; blank/null falls back to the manifest value.
        val clientId = call.argument<String>("clientId")
        val loginType = when (call.argument<String>("strategy")) {
            "webOnly" -> LoginType.CHROME_TAB
            else -> LoginType.NATIVE
        }
        pendingResult = result
        nativeLog("debug", "launching sign-in (loginType=$loginType)")
        try {
            launcher!!.launch(YandexAuthLoginOptions(loginType, clientId))
        } catch (e: Throwable) {
            pendingResult = null
            nativeLog("error", "launcher.launch failed: ${e.message}")
            result.error("SDK_ERROR", e.message ?: "launcher.launch failed", null)
        }
    }

    private fun deliverResult(authResult: YandexAuthResult) {
        val pending = pendingResult ?: return
        pendingResult = null
        when (authResult) {
            is YandexAuthResult.Success -> {
                nativeLog("debug", "sign-in succeeded (expiresIn=${authResult.token.expiresIn})")
                pending.success(
                    mapOf(
                        "token" to authResult.token.value,
                        "expiresIn" to authResult.token.expiresIn,
                    )
                )
            }
            is YandexAuthResult.Failure -> {
                nativeLog("error", "sign-in failed: ${authResult.exception.message}")
                pending.error(
                    "SDK_ERROR",
                    authResult.exception.message ?: "Yandex auth failed",
                    (authResult.exception as? YandexAuthException)?.errors?.joinToString(),
                )
            }
            YandexAuthResult.Cancelled -> {
                nativeLog("info", "sign-in cancelled by user")
                pending.error("CANCELLED", "User cancelled", null)
            }
        }
    }

    /**
     * Mirrors a native diagnostic event into the Dart-side `YandexLoginSdk.onLog`
     * hook. Gated by the `nativeLogging` flag the Dart layer sends with each
     * signIn (true only while a log handler is installed).
     */
    private fun nativeLog(level: String, message: String) {
        if (!loggingEnabled) return
        channel.invokeMethod("log", mapOf("level" to level, "message" to message))
    }

    private companion object {
        const val CHANNEL = "yandex_login_sdk"
        const val REGISTRY_KEY = "yandex_login_sdk:auth"
    }
}

File: ios/yandex_login_sdk/Sources/yandex_login_sdk/YandexLoginSdkPlugin.swift
import AuthenticationServices
import Flutter
import UIKit
import YandexLoginSDK

/// Flutter plugin wrapping the official Yandex `YandexLoginSDK` (iOS).
///
/// Apps that use the modern `UISceneDelegate` lifecycle (default for projects
/// created with Flutter 3+) must forward URL callbacks from their
/// `SceneDelegate.scene(_:openURLContexts:)` to
/// `YandexLoginSdkPlugin.handle(openURL:)`. AppDelegate-only apps don't need
/// any extra wiring — the plugin registers itself as a `FlutterApplicationLifeCycleDelegate`.
public class YandexLoginSdkPlugin: NSObject, FlutterPlugin {

    private var pendingResult: FlutterResult?
    private var activatedClientId: String?
    private var channel: FlutterMethodChannel?
    private var loggingEnabled = false

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "yandex_login_sdk",
            binaryMessenger: registrar.messenger()
        )
        let instance = YandexLoginSdkPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)
        YandexLoginSDK.shared.add(observer: instance)
    }

    /// Mirrors a native diagnostic event into the Dart-side
    /// `YandexLoginSdk.onLog` hook. Gated by the `nativeLogging` flag the Dart
    /// layer sends with each signIn (true only while a log handler is
    /// installed).
    private func nativeLog(_ level: String, _ message: String) {
        guard loggingEnabled else { return }
        channel?.invokeMethod("log", arguments: ["level": level, "message": message])
    }

    /// Forwards a URL callback to the underlying Yandex SDK. Call this from
    /// your `SceneDelegate.scene(_:openURLContexts:)`:
    ///
    /// ```swift
    /// func scene(_ scene: UIScene, openURLContexts URLContexts: Set<UIOpenURLContext>) {
    ///     for ctx in URLContexts {
    ///         _ = YandexLoginSdkPlugin.handle(openURL: ctx.url)
    ///     }
    /// }
    /// ```
    @discardableResult
    public static func handle(openURL url: URL) -> Bool {
        return YandexLoginSDK.shared.tryHandleOpenURL(url)
    }

    /// Forwards a Universal Link callback to the underlying Yandex SDK.
    @discardableResult
    public static func handle(continue userActivity: NSUserActivity) -> Bool {
        return YandexLoginSDK.shared.tryHandleUserActivity(userActivity)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "signIn":
            handleSignIn(call, result: result)
        case "signOut":
            handleSignOut(result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    /// Clears the SDK's locally-cached login state.
    ///
    /// `YandexLoginSDK.logout()` removes the stored token/JWT, the PKCE
    /// verifier and CSRF states from the Keychain. It is local-only: it does
    /// not revoke the token on Yandex's servers. Clearing the cache forces the
    /// next `authorize` to present interactive UI instead of returning the
    /// cached result.
    private func handleSignOut(_ result: @escaping FlutterResult) {
        do {
            try YandexLoginSDK.shared.logout()
            result(nil)
        } catch {
            result(FlutterError(code: "SDK_ERROR", message: error.localizedDescription, details: nil))
        }
    }

    private func handleSignIn(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard pendingResult == nil else {
            result(FlutterError(code: "BUSY", message: "Another sign-in is in progress", details: nil))
            return
        }
        guard
            let args = call.arguments as? [String: Any],
            let clientId = args["clientId"] as? String,
            !clientId.isEmpty
        else {
            result(FlutterError(code: "BAD_ARGS", message: "clientId is required", details: nil))
            return
        }
        let strategy: YandexLoginSDK.AuthorizationStrategy =
            (args["strategy"] as? String) == "webOnly" ? .webOnly : .default
        loggingEnabled = (args["nativeLogging"] as? Bool) ?? false
        do {
            // Re-activate when the app switches to a different OAuth client;
            // repeated activation with the same id is a cheap assignment.
            if activatedClientId != clientId {
                try YandexLoginSDK.shared.activate(with: clientId)
                activatedClientId = clientId
            }
            guard let rootVC = Self.topViewController() else {
                result(FlutterError(code: "NO_VC", message: "Root view controller not found", details: nil))
                return
            }
            pendingResult = result
            nativeLog("debug", "starting authorization (strategy=\(strategy))")
            try YandexLoginSDK.shared.authorize(
                with: rootVC,
                customValues: nil,
                authorizationStrategy: strategy
            )
        } catch {
            pendingResult = nil
            nativeLog("error", "authorize failed to start: \(error.localizedDescription)")
            result(FlutterError(code: "SDK_ERROR", message: error.localizedDescription, details: nil))
        }
    }

    private static func topViewController() -> UIViewController? {
        let scenes = UIApplication.shared.connectedScenes
        let windowScene = scenes.first(where: { $0.activationState == .foregroundActive }) as? UIWindowScene
            ?? scenes.first as? UIWindowScene
        let window = windowScene?.windows.first(where: { $0.isKeyWindow }) ?? windowScene?.windows.first
        var vc = window?.rootViewController
        while let presented = vc?.presentedViewController { vc = presented }
        return vc
    }

    /// `localizedDescription` is translated by the system, so cancellation is
    /// detected by error domain + code first; the English substring check is
    /// only a last resort for the SDK's own (non-localized) messages, such as
    /// `userClosedWebViewController`.
    private static func isCancellation(_ error: Error) -> Bool {
        if let asError = error as? ASWebAuthenticationSessionError,
           asError.code == .canceledLogin {
            return true
        }
        let nsError = error as NSError
        switch (nsError.domain, nsError.code) {
        case (ASWebAuthenticationSessionErrorDomain, ASWebAuthenticationSessionError.canceledLogin.rawValue),
             // SFAuthenticationErrorDomain / canceledLogin — the symbol is
             // deprecated, the literal avoids a build warning.
             ("com.apple.SafariServices.Authentication", 1),
             (NSCocoaErrorDomain, NSUserCancelledError):
            return true
        default:
            break
        }
        if let sdkError = error as? YandexLoginSDKError,
           sdkError.message.contains("closed the view controller") {
            return true
        }
        let desc = nsError.localizedDescription.lowercased()
        return desc.contains("cancel") || desc.contains("close")
    }

    // MARK: - UIApplicationDelegate (registered via addApplicationDelegate)

    public func application(
        _ application: UIApplication,
        open url: URL,
        options: [UIApplication.OpenURLOptionsKey: Any] = [:]
    ) -> Bool {
        return YandexLoginSDK.shared.tryHandleOpenURL(url)
    }

    public func application(
        _ application: UIApplication,
        continue userActivity: NSUserActivity,
        restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void
    ) -> Bool {
        return YandexLoginSDK.shared.tryHandleUserActivity(userActivity)
    }
}

extension YandexLoginSdkPlugin: YandexLoginSDKObserver {
    public func didFinishLogin(with result: Result<LoginResult, Error>) {
        guard let pending = pendingResult else { return }
        pendingResult = nil
        switch result {
        case .success(let loginResult):
            nativeLog("debug", "sign-in succeeded")
            pending([
                "token": loginResult.token,
                "jwt": loginResult.jwt,
            ])
        case .failure(let error):
            let cancelled = Self.isCancellation(error)
            nativeLog(
                cancelled ? "info" : "error",
                cancelled ? "sign-in cancelled by user" : "sign-in failed: \(error.localizedDescription)"
            )
            pending(FlutterError(
                code: cancelled ? "CANCELLED" : "SDK_ERROR",
                message: error.localizedDescription,
                details: nil
            ))
        }
    }
}
